lib/puppet/network/http/response.rb

```diff
@@ -5,7 +5,23 @@ def initialize(handler, response)
   end
 
   def respond_with(code, type, body)
-    @handler.set_content_type(@response, type)
+    format = Puppet::Network::FormatHandler.format_for(type)
+    mime = format.mime
+    charset = format.charset
+
+    if charset && body.is_a?(String)
+      # REMIND: not all charsets are valid ruby encodings, e.g. ISO-2022-KR
+      encoding = Encoding.find(charset)
+
+      if body.encoding != encoding
+        # REMIND this can raise if body contains invalid UTF-8
+        body.encode!(encoding)
```

Comment:
I believe raising if `encode!` fails is the correct thing to do. When sending `json`, the native JSON generator will ensure it emits valid utf-8 data, and will raise with invalid input:

```
irb(main):001:0> require 'json'
=> true
irb(main):002:0> "\355\274\267".to_json
JSON::GeneratorError: source sequence is illegal/malformed utf-8
...
```

So if `encode!` fails, then something went wrong between the time we rendered the result and here.

Note that `pson` will emit it as, but we don't attempt to encode `pson`:

```
irb(main):003:0> require 'puppet'
1=> true
irb(main):004:0> "\355\274\267".to_pson
=> "\"\xED\xBC\xB7\""
```